packages/flutter/lib/src/widgets/focus_manager.dart

```diff
@@ -68,330 +475,366 @@ class FocusNode extends ChangeNotifier {
+///   * [FocusScope], a widget that manages a [FocusScopeNode] and provides
+///     access to scope information and actions to its descendant widgets.
+///   * [FocusAttachment], a widget that connects a [FocusScopeNode] to the
+///     focus tree.
+///   * [FocusManager], a singleton that manages the focus and distributes key
+///     events to focused nodes.
+class FocusScopeNode extends FocusNode {
+  /// Creates a FocusScope node.
+  ///
+  /// All parameters are optional.
+  FocusScopeNode({
+    String debugLabel,
+    FocusOnKeyCallback onKey,
+  }) : super(debugLabel: debugLabel, onKey: onKey);
 
-  void _prepend(FocusScopeNode child) {
-    assert(child != this);
-    assert(child != _firstChild);
-    assert(child != _lastChild);
-    assert(child._parent == null);
-    assert(child._manager == null);
-    assert(child._nextSibling == null);
-    assert(child._previousSibling == null);
-    assert(() {
-      FocusScopeNode node = this;
-      while (node._parent != null)
-        node = node._parent;
-      assert(node != child); // indicates we are about to create a cycle
-      return true;
-    }());
-    child._parent = this;
-    child._nextSibling = _firstChild;
-    if (_firstChild != null)
-      _firstChild._previousSibling = child;
-    _firstChild = child;
-    _lastChild ??= child;
-    child._updateManager(_manager);
-  }
+  @override
+  FocusScopeNode get nearestScope => this;
 
-  void _updateManager(FocusManager manager) {
-    void update(FocusScopeNode child) {
-      if (child._manager == manager)
-        return;
-      child._manager = manager;
-      // We don't proactively null out the manager for FocusNodes because the
-      // manager holds the currently active focus node until the end of the
-      // microtask, even if that node is detached from the focus tree.
-      if (manager != null)
-        child._focus?._manager = manager;
-      child._visitChildren(update);
-    }
+  /// Returns true if this scope is the focused child of its parent scope.
+  bool get isFirstFocus => enclosingScope.focusedChild == this;
 
-    update(this);
+  /// Returns the child of this node that should receive focus if this scope
+  /// node receives focus.
+  ///
+  /// If [hasFocus] is true, then this points to the child of this node that is
+  /// currently focused.
+  ///
+  /// Returns null if there is no currently focused child.
+  FocusNode get focusedChild {
+    assert(_focusedChildren.isEmpty || _focusedChildren.last.enclosingScope == this,
+    'Focused child does not have the same idea of its enclosing scope as the scope does.');
+    return _focusedChildren.isNotEmpty ? _focusedChildren.last : null;
   }
 
-  void _visitChildren(void visitor(FocusScopeNode child)) {
-    FocusScopeNode child = _firstChild;
-    while (child != null) {
-      visitor(child);
-      child = child._nextSibling;
-    }
-  }
+  final List<FocusNode> _focusedChildren = <FocusNode>[];
```

Comment:
nit: the name makes it sound like all these children have focus where in reality it's just the order they where last focused in (and the last one is the one with actual focus)?

I don't have a good idea for a better name, but a quick comment may clear that up for the future reader.